Document/0x06h-Testing-Platform-Interaction.md

```diff
@@ -3,365 +3,2753 @@
+
+Please note that if the app uses [`openURL:options:completionHandler:`](https://developer.apple.com/documentation/uikit/uiapplication/1648685-openurl?language=objc "UIApplication openURL:options:completionHandler:") to open a universal link to the app's website, the link won't open in the app. As the call originates from the app, it won't be handled as a universal link.
+
+> From Apple Docs: When iOS launches your app after a user taps a universal link, you receive an `NSUserActivity` object with an `activityType` value of `NSUserActivityTypeBrowsingWeb`. The activity object’s `webpageURL` property contains the URL that the user is accessing. The webpage URL property always contains an HTTP or HTTPS URL, and you can use `NSURLComponents` APIs to manipulate the components of the URL. [...] To protect users’ privacy and security, you should not use HTTP when you need to transport data; instead, use a secure transport protocol such as HTTPS.
+
+From the note above we can highlight that:
+
+- the mentioned `NSUserActivity` object comes from the `continueUserActivity` parameter, as seen in the method above
+- the scheme of the `webpageURL` must be HTTP or HTTPS (any other scheme should throw an exception). The [`scheme` instance property](https://developer.apple.com/documentation/foundation/urlcomponents/1779624-scheme "URLComponents scheme") of `URLComponents` / `NSURLComponents` can be used to verify this
+
+If you don't have the original source code you can use Radare2 or Rabin2 to search the whole binary strings for the link receiver method:
+
+```bash
+$ rabin2 -zq Telegram\ X.app/Telegram\ X | grep restorationHan
+
+0x1000deea9 53 52 application:continueUserActivity:restorationHandler:
+```
+
+###### Checking the Data Handler Method
+
+You should check how the received data is validated. This can be also part of the method `application:continueUserActivity:restorationHandler:` method or might occur on a separate method being called from it. 
+
+Here's an example in Swift from the Telegram app:
+
+```swift
+func application(_ application: UIApplication, continue userActivity: NSUserActivity, 
+                restorationHandler: @escaping ([Any]?) -> Void) -> Bool {
+    ...
+    if userActivity.activityType == NSUserActivityTypeBrowsingWeb, let url = userActivity.webpageURL {
+        self.openUrl(url: url)
+    }
+    
+    return true
+}
+```
+
+You can see there that it verifies if the received activity is of type `NSUserActivityTypeBrowsingWeb` before obtaining the URL from `userActivity.webpageURL` and opening it. However, depending on the app, it would be a good idea to additionally apply a URL whitelist at this point.
```

Comment:
You can see there that it verifies if -> The example above verifies if

Reply:
In case it doesn't come back: Should we add here that this data can not be trusted?  We mention "You should check how the received data is validated." above, but don't really say why it needs to be validated.

Reply:
Very good point. Actually I see now that it is a bit vague. I've added more details about why this is important and how it should be verified.